js/classes/entities/midground/portal.js

```diff
@@ -1,18 +1,178 @@
+
+        // check if chad has come into portal range, then activate
+        if (this.chadDistance() < Portal.ACTIVATION_RADIUS) {
+            if (this.action === "inactive") {
+                this.action = "activating";
+                ASSET_MGR.playSFX(SFX.PORTAL_ACTIVATE.path, SFX.PORTAL_ACTIVATE.volume);
+            }
+            if (!ASSET_MGR.audioIsPlaying(SFX.PORTAL_IDLE.path)) {
+                ASSET_MGR.playMusic(SFX.PORTAL_IDLE.path, SFX.PORTAL_IDLE.volume);
+            }
+            
+        } else if (this.action === "active" && this.chadDistance() > Portal.ACTIVATION_RADIUS) {
+            ASSET_MGR.stopAudio(SFX.PORTAL_IDLE.path);
+        }
+
+        // activate portal after a delay
+        if (this.action === "activating") {
+            this.activationTimer -= GAME.clockTick;
+            if (this.activationTimer <= 0) {
+                this.action = "active";
+                ASSET_MGR.playMusic(SFX.PORTAL_IDLE.path, SFX.PORTAL_IDLE.volume);
+            }
+        }
+
+        // make portal perform a circular motion
+        this.pos.x += Math.cos(GAME.gameTime * 2) / 10;
+        this.pos.y += Math.sin(GAME.gameTime * 2) / 10;
 
+        if (this.action != "active") return;
+
+        
+        if (this.spawnTimer > 0) {
+            this.spawnTimer -= GAME.clockTick;
+        }
+        if (this.particleTimer > 0) {
+            this.particleTimer -= GAME.clockTick;
+        }
+
+        // release portal particles every 2 seconds
+        if (this.particleTimer <= 0) {
+            GAME.addEntity(new ParticleEffect(this.center, this.particleType));
+            this.particleTimer = Portal.PARTICLE_DELAY;
+        } 
+
+        // release enemies every 8 seconds
+        if (this.spawnTimer <= 0 && this.enemyStock.length > 0) {
+            for (let i = 0; i < this.spawnGroupSize; i++) {
+                GAME.addEntity(this.enemyStock.pop());
+            }
+            this.spawnTimer = Portal.SPAWN_DELAY;
+        }
     };
+    
+    draw() {
+        this.animations[this.action].drawFrame(Vector.worldToCanvasSpace(this.pos), Portal.SCALE);
+    };
+
+    loadAnimations() {
+        this.animations["inactive"] = new Animator(
+            Portal.SPRITESHEET,
+            new Vector(0, this.type * Portal.SIZE.y),
+            Portal.SIZE,
+            4, 0.5, true);
+
+        this.animations["activating"] = new Animator(
+            Portal.SPRITESHEET,
+            new Vector(Portal.SIZE.x * 4, this.type * Portal.SIZE.y),
+            Portal.SIZE,
+            1, 0.5, true);
+
+        this.animations["active"] = new Animator(
+            Portal.SPRITESHEET,
+            new Vector(Portal.SIZE.x * 5, this.type * Portal.SIZE.y),
+            Portal.SIZE,
+            2, 0.5, true);
+    }
+
+
+    /**
+     * Fills the portal with enemies.
+     * @param {Array} enemies The enemies to fill the portal with
+     */
+    fillWithEnemies(enemies) {
+        this.enemyStock.push(...enemies);
```

Comment:
Not aware of what ...enemies does, what does this syntax do? Does it fill a stack with the array following the `...`?